src/crystal/system/win32/thread.cr

```diff
@@ -16,42 +17,91 @@ class Thread
   property previous : Thread?
 
   def self.unsafe_each
-    @@threads.unsafe_each { |thread| yield thread }
+    threads.unsafe_each { |thread| yield thread }
   end
 
+  # Starts a new system thread.
+  def initialize(&@func : ->)
+    @th = uninitialized LibC::HANDLE
+
+    @th = GC.beginthreadex(
+      security: Pointer(Void).null,
+      stack_size: LibC::UInt.zero,
+      start_address: ->(data : Void*) { data.as(Thread).start; LibC::UInt.zero },
+      arglist: self.as(Void*),
+      initflag: LibC::UInt.zero,
+      thrdaddr: Pointer(LibC::UInt).null)
+
+    if @th.null?
+      raise RuntimeError.from_errno("_beginthreadex")
+    end
+  end
+
+  # Used once to initialize the thread object representing the main thread of
+  # the process (that already exists).
   def initialize
+    # `GetCurrentThread` returns a _constant_ and is only meaningful as an
+    # argument to Win32 APIs; to uniquely identify it we must duplicate the handle
+    @th = uninitialized LibC::HANDLE
+    cur_proc = LibC.GetCurrentProcess
+    LibC.DuplicateHandle(cur_proc, LibC.GetCurrentThread, cur_proc, pointerof(@th), 0, true, LibC::DUPLICATE_SAME_ACCESS)
+
+    @func = ->{}
     @main_fiber = Fiber.new(stack_address, self)
-    @@threads.push(self)
+
+    Thread.threads.push(self)
   end
 
-  @@current : Thread? = nil
+  private def detach
+    if @detached.compare_and_set(0, 1).last
+      yield
+    end
+  end
 
-  # Associates the Thread object to the running system thread.
-  protected def self.current=(@@current : Thread) : Thread
+  # Suspends the current thread until this thread terminates.
+  def join : Nil
+    detach do
+      if LibC.WaitForSingleObject(@th, LibC::INFINITE) != LibC::WAIT_OBJECT_0
+        @exception ||= RuntimeError.from_winerror("WaitForSingleObject")
+      end
+      if LibC.CloseHandle(@th) == 0
+        @exception ||= RuntimeError.from_winerror("CloseHandle")
+      end
+    end
+
+    if exception = @exception
+      raise exception
+    end
   end
 
+  @[ThreadLocal]
+  @@current : Thread?
+
   # Returns the Thread object associated to the running system thread.
   def self.current : Thread
-    @@current || raise "BUG: Thread.current returned NULL"
+    @@current ||= new
```

Comment:
Hmm. This can't be right. This sets up a new thread (including registering in `Thread.threads`) on `Thread.current`, but `Thread.current` is called **really** often and multiple times, for example whenever something looks up the current scheduler or fiber.